src/auth.py

```diff
@@ -2,35 +2,143 @@
     def wrapper(*args, **kwargs):
         base_handler_instance = args[0]
-        token = base_handler_instance.get_cluster_token()
-        request_token = base_handler_instance.request.headers.get(api_key.AUTH)
+        worker_id = kwargs.get(key.WORKER_ID_PARAM)
+        if base_handler_instance.get_worker_node(worker_id) is None:
+            # get_worker_node() internally aborts the request so return
+            return
+        else:
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+def validate_assignment(func):
+    def wrapper(*args, **kwargs):
+        base_handler_instance = args[0]
+        course_id = kwargs.get(key.COURSE_ID_PARAM)
+        assignment_name = kwargs.get(key.ASSIGNMENT_NAME_PARAM)
+        assignment_id = "{}/{}".format(course_id, assignment_name)
+        assignment = base_handler_instance.get_db().get_assignment_collection().find_one({key.ID: assignment_id})
+        if assignment is None:
+            base_handler_instance.abort(
+                {"message": "Course {} has not uploaded a config for assignment {}".format(course_id, assignment_name)},
+                BAD_REQUEST_CODE)
+            return
+        else:
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+def authenticate_cluster_token(func):
+    def wrapper(*args, **kwargs):
+        base_handler_instance = args[0]
+        expected_token = base_handler_instance.get_cluster_token()
+        request_token = base_handler_instance.request.headers.get(key.AUTH)
 
         if (request_token is None) or not request_token.startswith("Bearer ") or len(request_token.split(" ")) != 2:
             base_handler_instance.abort({"message": "Cluster token in wrong format. Expect format \'Bearer <token>\'"},
-                                        BAD_REQUEST_CODE)
-        elif token != request_token.split(" ")[1]:
+                                        UNAUTHORIZED_REQUEST_CODE)
+        elif expected_token != request_token.split(" ")[1]:
             base_handler_instance.abort({"message": "Not authorized. Wrong token."}, UNAUTHORIZED_REQUEST_CODE)
         else:
             return func(*args, **kwargs)
 
     return wrapper
+
+
+def authenticate_course(func):
+    def wrapper(*args, **kwargs):
+        base_handler_instance = args[0]
+
+        request_token = base_handler_instance.request.headers.get(key.AUTH)
+        if (request_token is None) or not request_token.startswith("Bearer ") or len(request_token.split(" ")) != 2:
+            base_handler_instance.abort({"message": "Cluster token in wrong format. Expect format \'Bearer <token>\'"},
+                                        UNAUTHORIZED_REQUEST_CODE)
+            return
+
+        request_token = request_token.split(" ")[1]
+
+        course_id = kwargs.get(key.COURSE_ID_PARAM)
+        course = base_handler_instance.get_course(course_id)
+        if course is None:
+            # get_course() internally aborts the request so return
```

Comment:
It was a very common functionality to abort when getting a course from the DB with the correct error message.  It led to a lot of repetitive code which is why I refactored all of these functionalities into the BaseApiHandler so that the client code can just `get_course` and if it returns None we can understand the course does not exist.

Reply:
I think this conflates getting a course from the DB and aborting, which is what I was trying to say in my original response. Two functionalities that should be separate, perhaps with a function that combines the two.